README.md
# kong-plugin-ubidots-aws-lambda

## :warning: This plugin is now maintained as part of [Kong](https://github.com/Kong/kong). Please open Issues and PRs in that repository

## :open_book: For documentation, please visit <https://docs.konghq.com/hub/kong-inc/ubidots-aws-lambda>


CHANGELOG.md
# Kong AWS Lambda plugin changelog

## Releasing new versions

- update changelog below
- update rockspec version
- update version in `handler.lua`
- commit as `chore(*) release x.y.z`
- tag commit as `x.y.z`
- push commit and tags
- upload to luarocks; `luarocks upload kong-plugin-ubidots-aws-lambda-x.y.z-1.rockspec --api-key=abc...`
- test rockspec; `luarocks install kong-plugin-ubidots-aws-lambda`

## unreleased

- fix: handle multivalueheaders [#59](https://github.com/Kong/kong-plugin-ubidots-aws-lambda/pull/59)

## ubidots-aws-lambda 3.5.4 22-Mar-2021

- tests: just fix the test suite so that it works with kong repo too

## ubidots-aws-lambda 3.5.3 19-Mar-2021

- fix: respect `skip_large_bodies` config setting when using
  AWS API Gateway compatibility

## ubidots-aws-lambda 3.5.2 6-Nov-2020

- tests: just fix test suite for upcoming change to `ssl_cert` on Kong project

## ubidots-aws-lambda 3.5.1 6-Oct-2020

- fix: `skip_large_bodies` to honor config setting

## ubidots-aws-lambda 3.5.0 22-Sep-2020

- feat: adding support for 'isBase64Encoded' flag in Lambda function responses
- fix: respect `skip_large_bodies` config setting even when not using
  AWS API Gateway compatibility

## ubidots-aws-lambda 3.4.0 12-May-2020

- Change `luaossl` to `lua-resty-openssl`
- fix: do not validate region name against hardcoded list of regions
- feat: add `host` configuration to allow for custom Lambda endpoints

## ubidots-aws-lambda 3.3.0 17-Apr-2020

- Fix: when reusing the proxy based connection do not do the handshake again.
- revamped HTTP connect method that allows for connection reuse in scenarios
  with a proxy and/or ssl based connections

## ubidots-aws-lambda 3.2.1 3-Mar-2020

- Maintenance release for CI purposes

## ubidots-aws-lambda 3.2.0 11-Feb-2020

- Encrypt IAM access and secret keys when relevant

## ubidots-aws-lambda 3.1.0 6-Jan-2020

- fix: reduce notice-level message to debug, to reduce log noise
- feat: added 3 regions; eu-north-1, me-south-1, eu-west-3

## ubidots-aws-lambda 3.0.1 13-Nov-2019

- Remove the no-longer supported `run_on` field from plugin config schema

## ubidots-aws-lambda 3.0.0 2-Oct-2019

- Renamed from `liamp` to `ubidots-aws-lambda` to supersede the `ubidots-aws-lambda` plugin
  from Kong core.
- Note that this version number is just to indicate it supersedes the
  previously build in one. The effective change from version 2.0 (build in)
  of the ubidots-aws-lambda plugin to 3.0 (this) is the combined set of changes
  mentioned below for Liamp versions `0.1.0` and `0.2.0`.

## Liamp 0.2.0 23-Sep-2019

- chore: convert the plugin to the PDK and new DB (developed against Kong 1.x)

## Liamp 0.1.0

- feat: if no credentiuals are provided, the plugin will automatically fetch
  EC2 or ECS credentials and use the AWS IAM roles retrieved for accessing the
  Lambda.
- feat: new option `awsgateway_compatible` to make the serialized request
  compatible with the AWS gateway format, making the plugin a drop-in
  replacement
- feat: new option `skip_large_bodies` to enable really large bodies (that
  have been cached to disk) to also be sent to the Lambda. Use with care!
- feat: added the ability to connect to the Lambda through a proxy


spec/fixtures/ubidots-aws-lambda.lua
local helpers = require "spec.helpers"

local fixtures = {
  dns_mock = helpers.dns_mock.new(),
  http_mock = {
    lambda_plugin = [[
      server {
          server_name mock_aws_lambda;
          listen 10001 ssl;
> if ssl_cert[1] then
> for i = 1, #ssl_cert do
          ssl_certificate     $(ssl_cert[i]);
          ssl_certificate_key $(ssl_cert_key[i]);
> end
> else
          ssl_certificate ${{SSL_CERT}};
          ssl_certificate_key ${{SSL_CERT_KEY}};
> end
          ssl_protocols TLSv1.1 TLSv1.2 TLSv1.3;
          location ~ "/2015-03-31/functions/(?:[^/])*/invocations" {
              content_by_lua_block {
                local function x()
                  local function say(res, status)
                    ngx.header["x-amzn-RequestId"] = "foo"
                    if string.match(ngx.var.uri, "functionWithUnhandledError") then
                      ngx.header["X-Amz-Function-Error"] = "Unhandled"
                    end
                    ngx.status = status
                    if string.match(ngx.var.uri, "functionWithBadJSON") then
                      local badRes = "{\"foo\":\"bar\""
                      ngx.header["Content-Length"] = #badRes + 1
                      ngx.say(badRes)
                    elseif string.match(ngx.var.uri, "functionWithNoResponse") then
                      ngx.header["Content-Length"] = 0
                    elseif string.match(ngx.var.uri, "functionWithBase64EncodedResponse") then
                      ngx.say("{\"statusCode\": 200, \"body\": \"dGVzdA==\", \"isBase64Encoded\": true}")
                    elseif string.match(ngx.var.uri, "functionWithMultiValueHeadersResponse") then
                      ngx.say("{\"statusCode\": 200, \"headers\": { \"Age\": \"3600\"}, \"multiValueHeaders\": {\"Access-Control-Allow-Origin\": [\"site1.com\", \"site2.com\"]}}")
                    elseif string.match(ngx.var.uri, "functionEcho") then
                      require("spec.fixtures.mock_upstream").send_default_json_response()
                    elseif type(res) == 'string' then
                      ngx.header["Content-Length"] = #res + 1
                      ngx.say(res)
                    else
                      ngx.req.discard_body()
                      ngx.header['Content-Length'] = 0
                    end
                    ngx.exit(0)
                  end
                  ngx.sleep(.2) -- mock some network latency
                  local invocation_type = ngx.var.http_x_amz_invocation_type
                  if invocation_type == 'Event' then
                    say(nil, 202)
                  elseif invocation_type == 'DryRun' then
                    say(nil, 204)
                  end
                  local qargs = ngx.req.get_uri_args()
                  ngx.req.read_body()
                  local request_body = ngx.req.get_body_data()
                  if request_body == nil then
                    local body_file = ngx.req.get_body_file()
                    if body_file then
                      ngx.log(ngx.DEBUG, "reading file cached to disk: ",body_file)
                      local file = io.open(body_file, "rb")
                      request_body = file:read("*all")
                      file:close()
                    end
                  end
                  print(request_body)
                  local args = require("cjson").decode(request_body)
                  say(request_body, 200)
                end
                local ok, err = pcall(x)
                if not ok then
                  ngx.log(ngx.ERR, "Mock error: ", err)
                end
              }
          }
      }
    ]]
  },
}

fixtures.stream_mock = {
  lambda_proxy = [[
    server {
      listen 13128;
      content_by_lua_block {
        require("spec.fixtures.forward-proxy-server").connect()
      }
    }
  ]],
}

fixtures.dns_mock:A {
  name = "lambda.us-east-1.amazonaws.com",
  address = "127.0.0.1",
}

return fixtures

spec/fixtures/ubidots-aws-lambda-raw.lua
local helpers = require "spec.helpers"

local fixtures = {
  dns_mock = helpers.dns_mock.new(),
  http_mock = {
    lambda_plugin = [[
      server {
          server_name mock_aws_lambda;
          listen 10001 ssl;
> if ssl_cert[1] then
> for i = 1, #ssl_cert do
          ssl_certificate     $(ssl_cert[i]);
          ssl_certificate_key $(ssl_cert_key[i]);
> end
> else
          ssl_certificate ${{SSL_CERT}};
          ssl_certificate_key ${{SSL_CERT_KEY}};
> end
          ssl_protocols TLSv1.1 TLSv1.2 TLSv1.3;
          location ~ "/2015-03-31/functions/(?:[^/])*/invocations" {
              content_by_lua_block {
                local function x()
                  local function say(res, status)
                    ngx.header["x-amzn-RequestId"] = "foo"
                    if string.match(ngx.var.uri, "functionWithUnhandledError") then
                      ngx.header["X-Amz-Function-Error"] = "Unhandled"
                    end
                    ngx.status = status
                    if string.match(ngx.var.uri, "functionWithBadJSON") then
                      local badRes = "{\"foo\":\"bar\""
                      ngx.header["Content-Length"] = #badRes + 1
                      ngx.say(badRes)
                    elseif string.match(ngx.var.uri, "functionWithNoResponse") then
                      ngx.header["Content-Length"] = 0
                    elseif string.match(ngx.var.uri, "functionWithBase64EncodedResponse") then
                      ngx.say("{\"statusCode\": 200, \"body\": \"dGVzdA==\", \"isBase64Encoded\": true}")
                    elseif string.match(ngx.var.uri, "functionWithMultiValueHeadersResponse") then
                      ngx.say("{\"statusCode\": 200, \"headers\": { \"Age\": \"3600\"}, \"multiValueHeaders\": {\"Access-Control-Allow-Origin\": [\"site1.com\", \"site2.com\"]}}")
                    elseif string.match(ngx.var.uri, "functionEcho") then
                      require("spec.fixtures.mock_upstream").send_default_json_response()
                    elseif type(res) == 'string' then
                      ngx.header["Content-Length"] = #res + 1
                      ngx.say(res)
                    else
                      ngx.req.discard_body()
                      ngx.header['Content-Length'] = 0
                    end
                    ngx.exit(0)
                  end
                  ngx.sleep(.2) -- mock some network latency
                  local invocation_type = ngx.var.http_x_amz_invocation_type
                  if invocation_type == 'Event' then
                    say(nil, 202)
                  elseif invocation_type == 'DryRun' then
                    say(nil, 204)
                  end
                  local qargs = ngx.req.get_uri_args()
                  ngx.req.read_body()
                  local request_body = ngx.req.get_body_data()
                  if request_body == nil then
                    local body_file = ngx.req.get_body_file()
                    if body_file then
                      ngx.log(ngx.DEBUG, "reading file cached to disk: ",body_file)
                      local file = io.open(body_file, "rb")
                      request_body = file:read("*all")
                      file:close()
                    end
                  end
                  print(request_body)
                  local body_decoded = require("cjson").decode(request_body)
                  body_decoded["status_code"] = 201
                  local result_body = {test = "value"}
                  result_body["body"] = body_decoded["body"]
                  result_body["headers"] = body_decoded["headers"]
                  result_body["path"] = body_decoded["path"]
                  body_decoded["body"] = require("cjson").encode(result_body)
                  body_decoded["headers"] = {test = "x"}
                  body_decoded["headers"]["Content-Type"] = "application/json"
                  say(require("cjson").encode(body_decoded), 200)
                end
                local ok, err = pcall(x)
                if not ok then
                  ngx.log(ngx.ERR, "Mock error: ", err)
                end
              }
          }
      }
    ]]
  },
}

fixtures.stream_mock = {
  lambda_proxy = [[
    server {
      listen 13128;
      content_by_lua_block {
        require("spec.fixtures.forward-proxy-server").connect()
      }
    }
  ]],
}

fixtures.dns_mock:A {
  name = "lambda.us-east-1.amazonaws.com",
  address = "127.0.0.1",
}

return fixtures

spec/plugins/ubidots-aws-lambda/02-schema_spec.lua
local schema_def = require "kong.plugins.ubidots-aws-lambda.schema"
local v = require("spec.helpers").validate_plugin_config_schema


describe("Plugin: AWS Lambda (schema)", function()
  it("accepts nil Unhandled Response Status Code", function()
    local ok, err = v({
      unhandled_status = nil,
      aws_region = "us-east-1",
      function_name = "my-function"
    }, schema_def)

    assert.is_nil(err)
    assert.truthy(ok)
  end)

  it("accepts correct Unhandled Response Status Code", function()
    local ok, err = v({
      unhandled_status = 412,
      aws_region = "us-east-1",
      function_name = "my-function"
    }, schema_def)

    assert.is_nil(err)
    assert.truthy(ok)
  end)

  it("errors with Unhandled Response Status Code less than 100", function()
    local ok, err = v({
      unhandled_status = 99,
      aws_region = "us-east-1",
      function_name = "my-function"
    }, schema_def)

    assert.equal("value should be between 100 and 999", err.config.unhandled_status)
    assert.falsy(ok)
  end)

  it("errors with Unhandled Response Status Code greater than 999", function()
    local ok, err = v({
      unhandled_status = 1000,
      aws_region = "us-east-1",
      function_name = "my-function"
    }, schema_def)

    assert.equal("value should be between 100 and 999", err.config.unhandled_status)
    assert.falsy(ok)
  end)

  it("accepts with neither aws_key nor aws_secret", function()
    local ok, err = v({
      aws_region = "us-east-1",
      function_name = "my-function"
    }, schema_def)

    assert.is_nil(err)
    assert.truthy(ok)
  end)

  it("errors with aws_secret but without aws_key", function()
    local ok, err = v({
      aws_secret = "xx",
      aws_region = "us-east-1",
      function_name = "my-function"
    }, schema_def)

    assert.equal("all or none of these fields must be set: 'config.aws_key', 'config.aws_secret'", err["@entity"][1])
    assert.falsy(ok)
  end)

  it("errors without aws_secret but with aws_key", function()
    local ok, err = v({
      aws_key = "xx",
      aws_region = "us-east-1",
      function_name = "my-function"
    }, schema_def)

    assert.equal("all or none of these fields must be set: 'config.aws_key', 'config.aws_secret'", err["@entity"][1])
    assert.falsy(ok)
  end)

  it("errors with a non-http proxy_url", function()
    for _, scheme in ipairs({"https", "ftp", "wss"}) do
      local ok, err = v({
        proxy_url = scheme .. "://squid:3128",
        aws_region = "us-east-1",
        function_name = "my-function"
      }, schema_def)

      assert.not_nil(err)
      assert.falsy(ok)
      assert.equals("proxy_url scheme must be http", err["@entity"][1])
    end
  end)

  it("accepts a host", function()
    local ok, err = v({
      host = "my.lambda.host",
      function_name = "my-function"
    }, schema_def)

    assert.is_nil(err)
    assert.truthy(ok)
  end)

  it("does not error if none of aws_region and host are passed (tries to autodetect on runtime)", function()
    local ok, err = v({
      function_name = "my-function"
    }, schema_def)

    assert.is_nil(err)
    assert.truthy(ok)
  end)

  it("errors if both of aws_region and host are passed", function()
    local ok, err = v({
      host = "my.lambda.host",
      aws_region = "us-east-1",
      function_name = "my-function"
    }, schema_def)

    assert.equal("only one or none of these fields must be set: 'config.aws_region', 'config.host'", err["@entity"][1])
    assert.falsy(ok)
  end)
end)


kong/plugins/ubidots-aws-lambda/schema.lua
local typedefs = require "kong.db.schema.typedefs"

return {
  name = "ubidots-aws-lambda",
  fields = {
    { protocols = typedefs.protocols_http },
    { config = {
      type = "record",
      fields = {
        { timeout = {
          type = "number",
          required = true,
          default = 60000,
        } },
        { keepalive = {
          type = "number",
          required = true,
          default = 60000,
        } },
        { aws_key = {
          type = "string",
          encrypted = true, -- Kong Enterprise-exclusive feature, does nothing in Kong CE
          referenceable = true,
        } },
        { aws_secret = {
          type = "string",
          encrypted = true, -- Kong Enterprise-exclusive feature, does nothing in Kong CE
          referenceable = true,
        } },
        { aws_region = typedefs.host },
        { function_name = {
          type = "string",
          required = false,
        } },
        { qualifier = {
          type = "string",
        } },
        { invocation_type = {
          type = "string",
          required = true,
          default = "RequestResponse",
          one_of = { "RequestResponse", "Event", "DryRun" }
        } },
        { log_type = {
          type = "string",
          required = true,
          default = "Tail",
          one_of = { "Tail", "None" }
        } },
        { host = typedefs.host },
        { port = typedefs.port { default = 443 }, },
        { unhandled_status = {
          type = "integer",
          between = { 100, 999 },
        } },
        { forward_request_method = {
          type = "boolean",
          default = false,
        } },
        { forward_request_uri = {
          type = "boolean",
          default = false,
        } },
        { forward_request_headers = {
          type = "boolean",
          default = false,
        } },
        { forward_request_body = {
          type = "boolean",
          default = false,
        } },
        { is_proxy_integration = {
          type = "boolean",
          default = false,
        } },
        { awsgateway_compatible = {
          type = "boolean",
          default = false,
        } },
        { proxy_url = typedefs.url },
        { skip_large_bodies = {
          type = "boolean",
          default = true,
        } },
        { base64_encode_body = {
          type = "boolean",
          default = true,
        } },
        { raw_function    = {
          type = "boolean",
          default  = false,
        } },
        { environment     = {
          type = "string",
          default  = "{}",
        } },
        { parameters      = {
          type = "string",
          default  = "{}",
        } },
        { auth_token      = {
          type = "string",
          required = false
        } },
        { methods         = {
              type = "array",
              default  = { "POST" },
              elements = {
                type = "string",
                one_of = { "HEAD", "GET", "POST", "PATCH", "PUT" }
              }
            }
        },
        { allowed_origins = {
              type = "array",
              default  = {},
              elements = {
                type = "string",
              }
            }
        },
      }
    },
  } },
  entity_checks = {
    { mutually_required = { "config.aws_key", "config.aws_secret" } },
    { mutually_exclusive = { "config.aws_region", "config.host" } },
    { custom_entity_check = {
        field_sources = { "config.proxy_url" },
        fn = function(entity)
          local proxy_url = entity.config and entity.config.proxy_url

          if type(proxy_url) == "string" then
            local scheme = proxy_url:match("^([^:]+)://")

            if scheme and scheme ~= "http" then
              return nil, "proxy_url scheme must be http"
            end
          end

          return true
        end,
      }
    },
  }
}

kong/plugins/ubidots-aws-lambda/aws-serializer.lua
-- serializer to wrap the current request into the Amazon API gateway
-- format as described here:
-- https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-input-format

local request_util = require "kong.plugins.ubidots-aws-lambda.request-util"

local EMPTY = {}

local ngx_req_get_headers  = ngx.req.get_headers
local ngx_req_get_uri_args = ngx.req.get_uri_args
local ngx_encode_base64    = ngx.encode_base64

return function(ctx, config)
  ctx = ctx or ngx.ctx
  local var = ngx.var

  -- prepare headers
  local headers = ngx_req_get_headers()
  local multiValueHeaders = {}
  for hname, hvalue in pairs(headers) do
    if type(hvalue) == "table" then
      -- multi value
      multiValueHeaders[hname] = hvalue
      headers[hname] = hvalue[1]

    else
      -- single value
      multiValueHeaders[hname] = { hvalue }
    end
  end

  -- prepare url-captures/path-parameters
  local pathParameters = {}
  for name, value in pairs(ctx.router_matches.uri_captures or EMPTY) do
    if type(name) == "string" then  -- skip numerical indices, only named
      pathParameters[name] = value
    end
  end

  -- query parameters
  local queryStringParameters = ngx_req_get_uri_args()
  local multiValueQueryStringParameters = {}
  for qname, qvalue in pairs(queryStringParameters) do
    if type(qvalue) == "table" then
      -- multi value
      multiValueQueryStringParameters[qname] = qvalue
      queryStringParameters[qname] = qvalue[1]

    else
      -- single value
      multiValueQueryStringParameters[qname] = { qvalue }
    end
  end

  -- prepare body
  local body, isBase64Encoded
  local skip_large_bodies = true
  local base64_encode_body = true

  if config then
    if config.skip_large_bodies ~= nil then
      skip_large_bodies = config.skip_large_bodies
    end

    if config.base64_encode_body ~= nil then
      base64_encode_body = config.base64_encode_body
    end
  end

  do
    body = request_util.read_request_body(skip_large_bodies)
    if body ~= "" and base64_encode_body then
      body = ngx_encode_base64(body)
      isBase64Encoded = true
    else
      isBase64Encoded = false
    end
  end

  -- prepare path
  local path = var.request_uri:match("^([^%?]+)")  -- strip any query args

  local request = {
    resource                        = ctx.router_matches.uri,
    path                            = path,
    httpMethod                      = var.request_method,
    headers                         = headers,
    multiValueHeaders               = multiValueHeaders,
    pathParameters                  = pathParameters,
    queryStringParameters           = queryStringParameters,
    multiValueQueryStringParameters = multiValueQueryStringParameters,
    body                            = body,
    isBase64Encoded                 = isBase64Encoded,
  }

  return request
end


kong/plugins/ubidots-aws-lambda/handler.lua
-- Copyright (C) Kong Inc.

local aws_v4 = require "kong.plugins.ubidots-aws-lambda.v4"
local aws_serializer = require "kong.plugins.ubidots-aws-lambda.aws-serializer"
local http = require "resty.http"
local cjson = require "cjson.safe"
local meta = require "kong.meta"
local constants = require "kong.constants"
local request_util = require "kong.plugins.ubidots-aws-lambda.request-util"


local VIA_HEADER = constants.HEADERS.VIA
local VIA_HEADER_VALUE = meta._NAME .. "/" .. meta._VERSION
local IAM_CREDENTIALS_CACHE_KEY = "plugin.ubidots-aws-lambda.iam_role_temp_creds"
local AWS_PORT = 443
local AWS_REGION do
  AWS_REGION = os.getenv("AWS_REGION") or os.getenv("AWS_DEFAULT_REGION")
end


local fetch_credentials do
  local credential_sources = {
    require "kong.plugins.ubidots-aws-lambda.iam-ecs-credentials",
    -- The EC2 one will always return `configured == true`, so must be the last!
    require "kong.plugins.ubidots-aws-lambda.iam-ec2-credentials",
  }

  for _, credential_source in ipairs(credential_sources) do
    if credential_source.configured then
      fetch_credentials = credential_source.fetchCredentials
      break
    end
  end
end


local ngx_encode_base64 = ngx.encode_base64
local ngx_decode_base64 = ngx.decode_base64
local ngx_update_time = ngx.update_time
local tostring = tostring
local tonumber = tonumber
local ngx_now = ngx.now
local ngx_var = ngx.var
local error = error
local pairs = pairs
local kong = kong
local type = type
local fmt = string.format


local raw_content_types = {
  ["text/plain"] = true,
  ["text/html"] = true,
  ["application/xml"] = true,
  ["text/xml"] = true,
  ["application/soap+xml"] = true,
}


local function get_now()
  ngx_update_time()
  return ngx_now() * 1000 -- time is kept in seconds with millisecond resolution.
end


--[[
  Response format should be
  {
      "statusCode": httpStatusCode,
      "headers": { "headerName": "headerValue", ... },
      "body": "..."
  }
--]]
local function validate_custom_response(response)
  if type(response.statusCode) ~= "number" then
    return nil, "statusCode must be a number"
  end

  if response.headers ~= nil and type(response.headers) ~= "table" then
    return nil, "headers must be a table"
  end

  if response.body ~= nil and type(response.body) ~= "string" then
    return nil, "body must be a string"
  end

  return true
end


local function get_blocking_invocation_type(blocking, invocation_type)
  if blocking == "true" or blocking == true then
    return "RequestResponse"
  elseif blocking == "false" or blocking == false then
    return "Event"
  else
    return invocation_type
  end
end

local function extract_raw_function_response(content, status_code)
  local response_json = cjson.decode(content)
  if response_json == nil then
    return {status_code = status_code}
  end
  local headers = response_json.headers or {}
  local body = response_json.body or "{}"
  headers["Content-Length"] = #body
  return {
    status_code = response_json.status_code,
    body = body,
    headers = headers
  }
end

local function extract_proxy_response(content)
  local serialized_content, err = cjson.decode(content)
  if not serialized_content then
    return nil, err
  end

  local ok, err = validate_custom_response(serialized_content)
  if not ok then
    return nil, err
  end

  local headers = serialized_content.headers or {}
  local body = serialized_content.body or ""
  local isBase64Encoded = serialized_content.isBase64Encoded or false
  if isBase64Encoded then
    body = ngx_decode_base64(body)
  end

  local multiValueHeaders = serialized_content.multiValueHeaders
  if multiValueHeaders then
    for header, values in pairs(multiValueHeaders) do
      headers[header] = values
    end
  end

  headers["Content-Length"] = #body

  return {
    status_code = tonumber(serialized_content.statusCode),
    body = body,
    headers = headers,
  }
end

local function is_dictionary_body(body)
  return type(body) == "table" and body[1] == nil
end

local AWSLambdaHandler = {}


function AWSLambdaHandler:access(conf)
  local upstream_body = kong.table.new(0, 6)
  local ctx = ngx.ctx

  if conf.raw_function then
    upstream_body = {}
  elseif conf.awsgateway_compatible then
    upstream_body = aws_serializer(ctx, conf)

  elseif conf.forward_request_body or
    conf.forward_request_headers or
    conf.forward_request_method or
    conf.forward_request_uri then

    -- new behavior to forward request method, body, uri and their args
    if conf.forward_request_method then
      upstream_body.request_method = kong.request.get_method()
    end

    if conf.forward_request_headers then
      upstream_body.request_headers = kong.request.get_headers()
    end

    if conf.forward_request_uri then
      upstream_body.request_uri = kong.request.get_path_with_query()
      upstream_body.request_uri_args = kong.request.get_query()
    end

    if conf.forward_request_body then
      local content_type = kong.request.get_header("content-type")
      local body_raw = request_util.read_request_body(conf.skip_large_bodies)
      local body_args, err = kong.request.get_body()
      if err and err:match("content type") then
        body_args = {}
        if not raw_content_types[content_type] and conf.base64_encode_body then
          -- don't know what this body MIME type is, base64 it just in case
          body_raw = ngx_encode_base64(body_raw)
          upstream_body.request_body_base64 = true
        end
      end

      upstream_body.request_body      = body_raw
      upstream_body.request_body_args = body_args
    end

  else
    -- backwards compatible upstream body for configurations not specifying
    -- `forward_request_*` values
    local body_args = kong.request.get_body()
    upstream_body = kong.table.merge(kong.request.get_query(), body_args)
  end
  local blocking = nil
  if is_dictionary_body(upstream_body) then
    blocking = (upstream_body['_blocking'])
    upstream_body["_blocking"] = nil
  end
  if conf.auth_token ~= nil and is_dictionary_body(upstream_body) then
    upstream_body['_auth_token'] = conf.auth_token
  end
  if conf.environment ~= nil and is_dictionary_body(upstream_body) then
    upstream_body['_environment'] = conf.environment
  end
  if conf.parameters ~= nil and is_dictionary_body(upstream_body) then
    upstream_body['_parameters'] = conf.parameters
  end
  if conf.raw_function then
    upstream_body['headers'] = kong.request.get_headers()
    upstream_body['body'] = request_util.read_request_body(conf.skip_large_bodies)
    upstream_body['path'] = kong.request.get_path_with_query()
  end
  local upstream_body_json, err = cjson.encode(upstream_body)
  if not upstream_body_json then
    kong.log.err("could not JSON encode upstream body",
                 " to forward request values: ", err)
  end

  local region = conf.aws_region or AWS_REGION
  local host = conf.host

  if not region and not host then
    return error("no region or host specified")
  end

  if not host then
    host = fmt("lambda.%s.amazonaws.com", region)
  end

  local path = fmt("/2015-03-31/functions/%s/invocations", conf.function_name)
  local port = conf.port or AWS_PORT

  local opts = {
    region = region,
    service = "lambda",
    method = "POST",
    headers = {
      ["X-Amz-Target"] = "invoke",
      ["X-Amz-Invocation-Type"] = get_blocking_invocation_type(blocking, conf.invocation_type),
      ["X-Amz-Log-Type"] = conf.log_type,
      ["Content-Type"] = "application/x-amz-json-1.1",
      ["Content-Length"] = upstream_body_json and tostring(#upstream_body_json),
    },
    body = upstream_body_json,
    path = path,
    host = host,
    port = port,
    query = conf.qualifier and "Qualifier=" .. conf.qualifier
  }

  if not conf.aws_key then
    -- no credentials provided, so try the IAM metadata service
    local iam_role_credentials = kong.cache:get(
      IAM_CREDENTIALS_CACHE_KEY,
      nil,
      fetch_credentials
    )

    if not iam_role_credentials then
      return kong.response.error(500)
    end

    opts.access_key = iam_role_credentials.access_key
    opts.secret_key = iam_role_credentials.secret_key
    opts.headers["X-Amz-Security-Token"] = iam_role_credentials.session_token

  else
    opts.access_key = conf.aws_key
    opts.secret_key = conf.aws_secret
  end

  local request
  request, err = aws_v4(opts)
  if err then
    return error(err)
  end

  local uri = port and fmt("https://%s:%d", host, port)
                    or fmt("https://%s", host)

  local proxy_opts
  if conf.proxy_url then
    -- lua-resty-http uses the request scheme to determine which of
    -- http_proxy/https_proxy it will use, and from this plugin's POV, the
    -- request scheme is always https
    proxy_opts = { https_proxy = conf.proxy_url }
  end

  -- Trigger request
  local client = http.new()
  client:set_timeout(conf.timeout)
  local kong_wait_time_start = get_now()
  local res, err = client:request_uri(uri, {
    method = "POST",
    path = request.url,
    body = request.body,
    headers = request.headers,
    ssl_verify = false,
    proxy_opts = proxy_opts,
    keepalive_timeout = conf.keepalive,
  })
  if not res then
    return error(err)
  end

  local content = res.body

  -- setting the latency here is a bit tricky, but because we are not
  -- actually proxying, it will not be overwritten
  ctx.KONG_WAITING_TIME = get_now() - kong_wait_time_start
  local headers = res.headers

  if ngx_var.http2 then
    headers["Connection"] = nil
    headers["Keep-Alive"] = nil
    headers["Proxy-Connection"] = nil
    headers["Upgrade"] = nil
    headers["Transfer-Encoding"] = nil
  end

  local status

  if conf.is_proxy_integration then
    local proxy_response, err = extract_proxy_response(content)
    if not proxy_response then
      kong.log.err(err)
      return kong.response.exit(502, { message = "Bad Gateway",
                                       error = "could not JSON decode Lambda " ..
                                         "function response: " .. err })
    end

    status = proxy_response.status_code
    headers = kong.table.merge(headers, proxy_response.headers)
    content = proxy_response.body
  end

  if not status then
    if conf.unhandled_status
      and headers["X-Amz-Function-Error"] == "Unhandled"
    then
      status = conf.unhandled_status

    else
      status = res.status
    end
  end

  headers = kong.table.merge(headers) -- create a copy of headers

  if kong.configuration.enabled_headers[VIA_HEADER] then
    headers[VIA_HEADER] = VIA_HEADER_VALUE
  end

  if conf.raw_function then
    local response_raw_function = extract_raw_function_response(content, status)
    status = response_raw_function.status_code
    headers = response_raw_function.headers
    content = response_raw_function.body
  end
  return kong.response.exit(status, content, headers)
end

AWSLambdaHandler.PRIORITY = 750
AWSLambdaHandler.VERSION = "3.6.3"

return AWSLambdaHandler